include/wil/network.h

```diff
@@ -0,0 +1,1338 @@
+    // SOCKADDR_INET
+    //   - a union of sockaddr_in and sockaddr_in6 -- i.e., large enough to contain any TCPIP IPv4 or IPV6 address
+    // in_addr / IN_ADDR
+    //   - the raw address portion of a sockaddr_in
+    // in6_addr / IN6_ADDR
+    //   - the raw address portion of a sockaddr_in6
+    //
+    // SOCKET_ADDRESS
+    //   - not a derived sockaddr* type
+    //   - a structure containing both a sockaddr* and its length fields, returned from some networking functions
+    //
+
+    // declaring char-arrays large enough for any IPv4 or IPv6 address + optional fields
+    static_assert(INET6_ADDRSTRLEN > INET_ADDRSTRLEN);
+    typedef WCHAR socket_address_wstring[INET6_ADDRSTRLEN];
+    typedef CHAR socket_address_string[INET6_ADDRSTRLEN];
+
+    class socket_address final
+    {
+    public:
+        constexpr socket_address() WI_NOEXCEPT = default;
+        explicit socket_address(ADDRESS_FAMILY family) WI_NOEXCEPT;
+        template <typename T>
+        explicit socket_address(_In_reads_bytes_(addr_size) const SOCKADDR* addr, T addr_size) WI_NOEXCEPT;
+        explicit socket_address(const SOCKADDR_IN*) WI_NOEXCEPT;
+        explicit socket_address(const SOCKADDR_IN6*) WI_NOEXCEPT;
+        explicit socket_address(const SOCKADDR_INET*) WI_NOEXCEPT;
+        explicit socket_address(const SOCKET_ADDRESS*) WI_NOEXCEPT;
+        explicit socket_address(const IN_ADDR*, unsigned short port = 0) WI_NOEXCEPT;
+        explicit socket_address(const IN6_ADDR*, unsigned short port = 0) WI_NOEXCEPT;
+#if defined(WIL_ENABLE_EXCEPTIONS)
+        explicit socket_address(PCWSTR, unsigned short port = 0);
+#endif
+
+        ~socket_address() = default;
+        socket_address(const socket_address&) WI_NOEXCEPT = default;
+        socket_address& operator=(const socket_address&) WI_NOEXCEPT = default;
+        socket_address(socket_address&&) WI_NOEXCEPT = default;
+        socket_address& operator=(socket_address&&) WI_NOEXCEPT = default;
+
+        bool operator==(const socket_address&) const WI_NOEXCEPT;
+        bool operator!=(const socket_address&) const WI_NOEXCEPT;
+        bool operator<(const socket_address&) const WI_NOEXCEPT;
+        bool operator>(const socket_address&) const WI_NOEXCEPT;
+
+        void swap(socket_address&) WI_NOEXCEPT;
+        void reset(ADDRESS_FAMILY family = AF_UNSPEC) WI_NOEXCEPT;
+
+        // set_sockaddr overwrites the entire sockaddr in the object (including address family)
+        template <typename T>
+        void set_sockaddr(_In_reads_bytes_(addr_size) const SOCKADDR* addr, T addr_size) WI_NOEXCEPT;
+        void set_sockaddr(const SOCKADDR_IN*) WI_NOEXCEPT;
+        void set_sockaddr(const SOCKADDR_IN6*) WI_NOEXCEPT;
+        void set_sockaddr(const SOCKADDR_INET*) WI_NOEXCEPT;
+        void set_sockaddr(const SOCKET_ADDRESS*) WI_NOEXCEPT;
+#if defined(WIL_ENABLE_EXCEPTIONS)
+        void set_sockaddr(SOCKET);
+        void set_sockaddr(PCWSTR);
+        void set_sockaddr(PCSTR);

```

Comment:
Why not also call these `reset` (to match the constructors)? AFAICT they all clear out `m_sockaddr`.

Reply:
Great idea.

Reply:
This helped fix related names too.